### PR TITLE
Core/fix/stash aspect

### PR DESCRIPTION
**Pull request by SandraBarmina**
Отрефакторила и поправила часть багов

### CODE REVIEW

File: page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/allure/ErrorHandler.java [hunk 1]
```diff
@@ -23,11 +23,9 @@ public static void attachScreenshot() {
         }
     }
 
-    public static void attachError(String title, Throwable throwable) {
+    public static void attachError(Throwable e) {
```
Comment (clicman) [inline page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/allure/ErrorHandler.java]: we need to supply a custom error title. so revert back signature. or create two methods

File: page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/aspects/CriticalStepAspect.java [hunk 2]
```diff
@@ -1,91 +1,86 @@
...
+    @Pointcut(value = "execution(* cucumber.runtime.snippets.SnippetGenerator.getSnippet(..)) && args(step,..) && if()")
+    public static boolean getSnippet(PickleStep step) {
+        return step.getText().startsWith(NON_CRITICAL);
     }
 
-    @Around(value = "addSignOfCritically(featurePath, step)", argNames = "joinPoint,featurePath,step")
+    @Around(value = "addSignOfCritically(featurePath, step)")
     public Object addSignOfCritically(ProceedingJoinPoint joinPoint, String featurePath, PickleStep step) throws Throwable {
-        PickleStepCustom newStep;
+        boolean isCritical = true;
```
Comment (clicman) [inline page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/aspects/CriticalStepAspect.java]: boolean isCritical = stepText.startsWith(NON_CRITICAL ? false : true;
    String stepText = stepText.startsWith(NON_CRITICAL ? stepText.replaceFirst("\\" + NON_CRITICAL, ""): step.getText();

File: page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/aspects/CriticalStepCheckAspect.java [hunk 3]
```diff
@@ -21,106 +21,107 @@
...
-
     @Pointcut("execution(* cucumber.runtime.StepDefinitionMatch.runStep(..))")
     public void runStep() {
     }
 
     @Pointcut("execution(* cucumber.runner.EventBus.send(..)) && args(event,..) && if()")
-    public static boolean sendStepFinished(Event event) {
-        return event instanceof TestStepFinished;
+    public static boolean sendStepFinished(TestStepFinished event) {
+        return !event.testStep.isHook();
```
Comment (clicman) [inline page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/aspects/CriticalStepCheckAspect.java]: Does it really handle instanceOf validation on the fly?

File: page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/aspects/CriticalStepCheckAspect.java [hunk 4]
```diff
@@ -21,106 +21,107 @@
...
+            } catch (Throwable e) {
+                attachError((PickleStepCustom) step, e);
             }
         } else {
             joinPoint.proceed();
         }
     }
 
+    private boolean isCritical(PickleStep step) {
+        if (step instanceof PickleStepCustom) {
```
Comment (clicman) [inline page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/aspects/CriticalStepCheckAspect.java]: return step instanceof PickleStepCustom ? ((PickleStepCustom) step).isCritical() : true;

File: page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/data/StashParser.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        Pattern stepDataPattern = Pattern.compile(PATH_PARSE_REGEX);
+        Matcher stepDataMatcher = stepDataPattern.matcher(stepText);
+        StringBuilder replacedValue = new StringBuilder(stepText);
+
+        for (Argument argument : testStep.getDefinitionArgument()) {
+            String argVal = argument.getVal();
+            int argOffset = argument.getOffset();
+
+            if (argVal != null) {
+                if (stepDataPattern.matcher(argVal).find() && stepDataMatcher.find()) {
```
Comment (clicman) [inline page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/data/StashParser.java]: Wrap if content to separate private method

File: page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/data/StashParser.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+    private static void replaceStepText(PickleStepCustom currentStep) throws IllegalAccessException {
+        currentStep.setText(StashParser.replaceDataPlaceholders(currentStep, currentStep.getText()));
+    }
+
+    private static void replacePickleArguments(PickleStepCustom currentStep) throws IllegalAccessException {
+        for (gherkin.pickles.Argument argument : currentStep.getArgument()) {
+            if (argument.getClass().equals(PickleTable.class)) {
+                replaceDataTable(currentStep, (PickleTable) argument);
+            } else {
```
Comment (clicman) [inline page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/data/StashParser.java]: else if?

File: page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/data/StashParser.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+                break;
+            }
+            replacedValue = replacedValue.replace(stepDataMatcher.start(), stepDataMatcher.end(), stashValue);
+            stepDataMatcher = stepDataPattern.matcher(replacedValue);
+        }
+        return replacedValue.toString();
+    }
+
+    private static String getStashValue(PickleStepCustom currentStep, String stashKey) {
+        Object stashValue = Stash.getValue(stashKey);
```
Comment (clicman) [inline page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/data/StashParser.java]: String result = stashValue != null && stashValue instanceof String ? (String) stashValue : null;
    String messageTemplate = stashValue == null && result == null ? "Stash value not found by key: " : "The value received by the key must be a string. Key: ";

    saveMessage(currentStep, message + stashKey);
    return result;

File: plugins/mobile-plugin/pom.xml [hunk 8]
```diff
@@ -18,7 +18,7 @@
   <name>mobile-plugin</name>
 
   <properties>
-    <skipTests>false</skipTests>
+    <skipTests>true</skipTests>
```
Comment (clicman) [inline plugins/mobile-plugin/pom.xml]: revert

File: page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/aspects/CriticalStepCheckAspect.java (hunk 3 above)
Comment (SandraBarmina) [reply page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/aspects/CriticalStepCheckAspect.java]: absolutely!

File: page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/aspects/CriticalStepAspect.java [hunk 9]
```diff
@@ -1,91 +1,79 @@
...
-                    }
-                } else {
-                    if (!isCritical) {
-                        arg = new ArgumentCustom(argument.getOffset() + NON_CRITICAL.length(), argument.getVal(), true);
-                    }
-                }
-                shiftedArguments.add(Optional.ofNullable(arg).orElse(argument));
+        List<Argument> shiftedArguments = new ArrayList<>();
+        for (Argument argument : arguments) {
+            Argument arg = argument;
```
Comment (kosteman) [inline page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/aspects/CriticalStepAspect.java]: А это для чего так делается?

File: page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/aspects/report/PrintParameters.java [hunk 10]
```diff
@@ -18,6 +18,7 @@
 import org.aspectj.lang.annotation.Aspect;
 import org.aspectj.lang.annotation.Before;
 import org.aspectj.lang.annotation.Pointcut;
+import ru.sbtqa.tag.pagefactory.optional.PickleStepCustom;
```
Comment (kosteman) [inline page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/aspects/report/PrintParameters.java]: импорт точно используется?

**Pull request merged**